Data/HashMap/Base.hs

```diff
@@ -751,12 +1083,131 @@ update f = alter (>>= f)
+alterF f !k !m = (<$> f mv) $ \fres ->
+  case fres of
+    Nothing -> delete' h k m
+    Just v' -> insert' h k v' m
+
+  where !h = hash k
+        mv = lookup' h k m
+{-# INLINABLE [0] alterF #-}
+
+#if MIN_VERSION_base(4,8,0)
+bogus :: a
+bogus = error "Data.HashMap.alterF internal error: hit bottom"
+
+bogus# :: (# #) -> (# a #)
+bogus# _ = error "Data.HashMap.alterF internal error: hit bottom#"
+
+{-# RULES
+"alterFWeird" forall f k m. alterF f k m = alterFWeird (f Nothing) (f (Just bogus)) f k m
+
+-- This rule covers the case where alterF is used to imitate 'insert'.
+-- Note that @x@ appears /twice/ on the LHS. That lets us know that
+-- the function will always produce the same value, and in particular
+-- cannot use the old value in any way.
+"alterFinsert" forall (f :: Maybe a -> Identity (Maybe a)) x k m.
+  alterFWeird (coerce (Just x)) (coerce (Just x)) f k m =
+    Identity (insert k x m)
+
+-- This rule handles the case where 'alterF' is used to do 'insertWith'-like
+-- things. Whenever possible, GHC will get rid of the Maybe nonsense for us.
+-- We delay this rule to stage 1 so alterFinsert has a chance to fire.
+"alterFinsertWith" [1] forall (f :: Maybe a -> Identity (Maybe a)) x y k m.
+  alterFWeird (coerce (Just x)) (coerce (Just y)) f k m =
+    Identity (insertModifying x (\mold -> case runIdentity (f (Just mold)) of
+                                            Nothing -> bogus# (# #)
+                                            Just new -> (# new #)) k m)
+
+-- Handle the case where someone uses 'alterF' instead of 'adjust'. This
+-- rule is kind of picky; it will only work if the function doesn't
+-- do anything between case matching on the Maybe and producing a result.
+"alterFadjust" forall (f :: Maybe a -> Identity (Maybe a)) _y k m.
+  alterFWeird (coerce Nothing) (coerce (Just _y)) f k m =
+    Identity (adjust# (\x -> case runIdentity (f (Just x)) of
+                               Just x' -> (# x' #)
+                               Nothing -> bogus# (# #)) k m)
+
+
+-- Here we know that f doesn't use its argument at all.
+"alterFdelete" forall (f :: Maybe a -> Identity (Maybe a)) k m.
+  alterFWeird (coerce Nothing) (coerce Nothing) f k m = Identity (delete k m)
+
+-- The simple specialization to Const; in this case we can look up
+-- the key without caring what position it's in.
+"alterFlookup" forall _ign1 _ign2 (f :: Maybe a -> Const r (Maybe a)) k m.
+  alterFWeird _ign1 _ign2 f k m = k `seq` m `seq` Const (getConst (f (lookup k m)))
+ #-}
+
+alterFWeird
+       :: (Functor f, Eq k, Hashable k)
+       => f (Maybe v)
+       -> f (Maybe v)
+       -> (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+alterFWeird _ _ f k m = alterFEager f k m
+{-# INLINE [0] alterFWeird #-}
+
+alterFEager :: (Functor f, Eq k, Hashable k)
+       => (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+alterFEager f !k m = (<$> f mv) $ \fres ->
+  case fres of
```

Comment:
Explain why we don't look up `k` in `m` on the outside, if indeed there's a good reason.

Reply:
On the outside meaning before this function is called and instead pass in `lookupRes`?

Reply:
I mean the reason we don't have

```haskell
alterFEager f k m = case lookupRecordBlah k m of ...

Reply:
No reason. It actually looks like if you do it the way you mentioned you'll save a case statement. Currently we have 3, I think that could be reduced to 2.

Reply:
There's an extra deletion thunk, I believe. I'm not sure we want that.